notebooks/07 - Crossfilter.ipynb

```diff
@@ -0,0 +1,407 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "hist_fig.update_layout(\n",
+    "    title = 'Earthquakes by magnitude',\n",
+    "    height = 350, \n",
+    "    xaxis = {'title': 'Magnitude'},\n",
+    "    yaxis = {'title': 'Count'},\n",
+    "    dragmode = 'select',\n",
+    "    hovermode = 'closest'\n",
+    ")\n",
+    "hist_fig"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7b42953b-5820-4c21-9c3f-a0cfbe675307",
+   "metadata": {},
+   "source": [
+    "You can now open the left side bar in the map and select **Filters** in the top nav menu:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "34e48a68-39d4-44e0-b6e1-2b562f3a7de4",
+   "metadata": {},
+   "source": [
+    "<img src=images/studio-filters-pane-2.jpg width=350>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "b72c25c9-3386-4d54-8d7a-984c125c7908",
+   "metadata": {},
+   "source": [
+    "Try changing the selection in the Plotly histogram by dragging (make sure you are using the **Box Select** tool). You should see the Magnitude filter update in the map when you change the selection in the Plotly histogram and vice versa:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f4f797db-0111-419d-b399-93fe14e43127",
+   "metadata": {},
+   "source": [
+    "<img src=images/crossfilter-480.gif width=480>"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8ea1685f-2f84-4423-9fa9-50883bf36cfd",
+   "metadata": {},
+   "source": [
+    "## Scatterplot"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "15dd9f89-4546-4203-b97b-368813184ac7",
+   "metadata": {},
+   "source": [
+    "Let's now do the same with a more sophisticated chart, a scatterplot:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "96ec1f4b-c654-43b0-b165-612c15f4554e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import plotly.graph_objects as go\n",
+    "scatter_fig = go.FigureWidget([\n",
+    "    go.Scattergl(\n",
+    "        x = df['Magnitude'],\n",
+    "        y = df['Depth'], \n",
+    "        marker = {'color': df['Depth'], 'size': df['Magnitude']}, \n",
+    "        mode = 'markers'\n",
+    "    )\n",
+    "])\n",
+    "scatter = scatter_fig.data[0]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "848e65a3-78d6-4363-9e19-e20ab60512b0",
+   "metadata": {},
+   "source": [
+    "Let's again add event handlers. This function will be called when the selection in the scatterplot is changed:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "10145222-b7c3-44c4-80dc-eca4391b746c",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "depth_filter = [df['Depth'].min(), df['Depth'].max()]\n",
+    "magnitude_filter = [df['Magnitude'].min(), df['Magnitude'].max()]\n",
+    "\n",
+    "def on_scatterplot_selection_change(trace, points, state):\n",
+    "    global magnitude_filter\n",
+    "    global depth_filter\n",
```

Comment:
It might be worth adding a markdown line or a code comment explaining that these are necessary. Seems easy to miss for a lot of users (I don't think `global` is a very commonly used Python keyword) and I presume it would break without them.